apps/src/music/player/MusicLibrary.ts

```diff
@@ -62,23 +70,40 @@ export default class MusicLibrary {
     return null;
   }
 
-  getFolderForPath(path: string): SoundFolder | null {
-    return this.groups[0].folders.find(folder => folder.path === path) || null;
+  // Given a sound ID (e.g. "pack1/sound1"), return the SoundFolder.
+  getFolderForSoundId(id: string): SoundFolder | null {
+    const lastSlashIndex = id.lastIndexOf('/');
+    const folderId = id.substring(0, lastSlashIndex);
+
+    return this.getFolderForFolderId(folderId);
+  }
+
+  // Given a folder ID (e.g. "pack1") return the SoundFolder.
+  getFolderForFolderId(folderId: string): SoundFolder | null {
+    return this.folders.find(folder => folder.id === folderId) || null;
   }
 
   // A progression step might specify a smaller set of allowed sounds.
   setAllowedSounds(allowedSounds: Sounds): void {
     this.allowedSounds = allowedSounds;
   }
 
+  generateSoundUrl(folder: SoundFolder, soundData: SoundData): string {
+    const baseUrl = soundData.restricted
+      ? baseAssetUrlRestricted
+      : getBaseAssetUrl();
+
+    const optionalSoundPath = soundData.path ? `${soundData.path}/` : '';
+    const url = `${baseUrl}${this.libraryJson.path}/${folder.path}/${optionalSoundPath}${soundData.src}.mp3`;
+    return url;
```

Comment:
nit: inline return?

Reply:
Hah, I think I do these for easier stepping and inspecting in the debugger, but can indeed tidy them up now.